test/hotspot/jtreg/sources/TestNoNULL.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8343802
+ * @summary Test prevent NULL backsliding in hotspot code and tests
+ * @run main TestNoNULL
+ */
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class TestNoNULL {
+    private static final Set<String> excludedSourceFiles = new HashSet<>();
+    private static final Set<String> excludedTestFiles = new HashSet<>();
+    private static final Set<String> excludedTestExtensions = Set.of(".c", ".java", ".jar", ".class", ".zip");
+    private static final Pattern NULL_PATTERN = Pattern.compile("\\bNULL\\b");
+    private static Path dir = Paths.get(System.getProperty("test.src"));
+    private static int errorCount = 0;
+
+    public static void main(String[] args) throws IOException {
+        int maxIter = 20;
+        while (maxIter-- > 0 && dir != null && !Files.exists(dir.resolve("src"))) {
+            dir = dir.getParent();
+        }
+
+        if (dir == null) {
+            throw new RuntimeException("Could not locate the 'src' directory within 20 parent directories.");
+        }
+
+        Path srcPath = dir.resolve("src").resolve("hotspot");
+        Path testPath = dir.resolve("test").resolve("hotspot");
+
+        initializeExcludedPaths(dir);
+
+        if (Files.exists(srcPath)) {
+            processFiles(srcPath, excludedSourceFiles, false);
+        }
+        processFiles(testPath, excludedTestFiles, true);
```

Comment:
I think it would clearer to pass not boolean flag, but Set `excludedTestExtensions` itself. And pass empty set for `srcPath`.
```
        if (Files.exists(srcPath)) {
            processFiles(srcPath, excludedSourceFiles, Set.of());
        }
        processFiles(testPath, excludedTestFiles, excludedTestExtensions);
```